### ISSUE TITLE
IllegalAccessException: class androidx.test.espresso.base.ThreadPoolExecutorExtractor$2 cannot access a member of class androidx.loader.content.ModernAsyncTask

### ISSUE THREAD

**Issue by charlesng** [robolectric/robolectric#6593]
### Description

Hello. 
I got the following exception in cmd after upgrade to 4.6.1. Before upgrade, no exception is thrown.

```
java.lang.RuntimeException: java.lang.IllegalAccessException: class androidx.test.espresso.base.ThreadPoolExecutorExtractor$2 cannot access a member of class androidx.loader.content.ModernAsyncTask with modifiers "public static final"
	at androidx.test.espresso.base.ThreadPoolExecutorExtractor.getCompatAsyncTaskThreadPool(ThreadPoolExecutorExtractor.java:4)
	at androidx.test.espresso.base.BaseLayerModule.provideCompatAsyncTaskMonitor(BaseLayerModule.java:1)
	at androidx.test.espresso.base.BaseLayerModule_ProvideCompatAsyncTaskMonitorFactory.provideCompatAsyncTaskMonitor(BaseLayerModule_ProvideCompatAsyncTaskMonitorFactory.java:1)
	at androidx.test.espresso.base.BaseLayerModule_ProvideCompatAsyncTaskMonitorFactory.get(BaseLayerModule_ProvideCompatAsyncTaskMonitorFactory.java:1)
	at androidx.test.espresso.base.BaseLayerModule_ProvideCompatAsyncTaskMonitorFactory.get(BaseLayerModule_ProvideCompatAsyncTaskMonitorFactory.java:2)
	at androidx.test.espresso.core.internal.deps.dagger.internal.DoubleCheck.get(DoubleCheck.java:6)
	at androidx.test.espresso.base.UiControllerImpl_Factory.get(UiControllerImpl_Factory.java:1)
	at androidx.test.espresso.base.UiControllerImpl_Factory.get(UiControllerImpl_Factory.java:2)
	at androidx.test.espresso.core.internal.deps.dagger.internal.DoubleCheck.get(DoubleCheck.java:6)
	at androidx.test.espresso.base.UiControllerModule_ProvideUiControllerFactory.get(UiControllerModule_ProvideUiControllerFactory.java:1)
	at androidx.test.espresso.base.UiControllerModule_ProvideUiControllerFactory.get(UiControllerModule_ProvideUiControllerFactory.java:2)
	at androidx.test.espresso.core.internal.deps.dagger.internal.DoubleCheck.get(DoubleCheck.java:6)
	at androidx.test.espresso.DaggerBaseLayerComponent$ViewInteractionComponentImpl.viewInteraction(DaggerBaseLayerComponent.java:1)
	at androidx.test.espresso.Espresso.onView(Espresso.java:1)
```

### Steps to Reproduce

- Upgrade to 4.6.1

### Robolectric & Android Version

- Robolectric : 4.6.1
- Espresso: 3.3.0
- Android Test Core: 1.3.0
- AGP: 4.1.3
- JDK11

**Comment by utzcoz** [comment]
Hi @charlesng , could you help to provide a minimal produced sample for this problem? Thanks.

**Comment by hoisie** [comment]
Ugh...

In https://github.com/robolectric/robolectric/pull/6188, we disabled instrumenting androidx.* packages by default. It's a nice performance improvement. But it seems that there's some logic in Espresso that was inadvertently relying on Robolectric's instrumentation to strip 'final' modifiers to be able to access final fields androidx.loader.content.ModernAsyncTask. To fix this issue temporarily, you can add:

```
@Config(
    instrumentedPackages = {
      // required to access final members on androidx.loader.content.ModernAsyncTask
      "androidx.loader.content"
    })
```

I'll reach out to the Espresso/AndroidX team to see if there's something that can be done here.

**Comment by hoisie** [comment]
This may be dead code in Espresso. The HEAD version of ModernAsyncTask does not even have a 'THREAD_POOL_EXECUTOR' field.

https://cs.android.com/androidx/platform/frameworks/support/+/androidx-main:loader/loader/src/main/java/androidx/loader/content/ModernAsyncTask.java

Maybe this logic can just be deleted in Espresso outright.

**Comment by hoisie** [comment]
Sent out https://github.com/android/android-test/pull/1000 to potentially fix this.

**Comment by calvarez-ov** [comment]
> To fix this issue temporarily, you can add:
> 
> ```
> @Config(
>     instrumentedPackages = {
>       // required to access final members on androidx.loader.content.ModernAsyncTask
>       "androidx.loader.content"
>     })
> ```

Thanks for this workaround! 🙏 

For info, another option is to add the following to `robolectric.properties`:

```
instrumentedPackages=androidx.loader.content
```

**Comment by charlesng** [comment]
> > To fix this issue temporarily, you can add:
> > ```
> > @Config(
> >     instrumentedPackages = {
> >       // required to access final members on androidx.loader.content.ModernAsyncTask
> >       "androidx.loader.content"
> >     })
> > ```
> 
> Thanks for this workaround! 🙏
> 
> For info, another option is to add the following to `robolectric.properties`:
> 
> ```
> instrumentedPackages=androidx.loader.content
> ```

This works for me! Thanks a lot for your guys help 🙇

**Comment by utzcoz** [comment]
> Sent out [android/android-test#1000](https://github.com/android/android-test/pull/1000) to potentially fix this.

Hi @charlesng and @calvarez-ov , above @hoisie 's PR is merged and released at `androidx-test-1.4.1-alpha01`, maybe you can try this version or later stable versions of `androidx-test` to check this problem again without above workaround.

**Comment by alwa** [comment]
> androidx-test-1.4.1-alpha01

Hi, I'm still getting the IllegalAccessException if removing the workaround after updating my dependencies to:

androidx-test-runner: 1.4.1-alpha02
androidx-test-monitor: 1.4.0-rc01
androidx-test-junit: 1.1.4-alpha02

**Comment by omesquita** [comment]
Hi guys! 
Does someone have an update for this issue?
I try to follow above susgestions but only add` instrumentedPackages=androidx.loader.content` the in robolectric.properties work. The application that i work have a lot of modules, it gets much more complex to maintain. 

Thanks a lot.

**Comment by nborodikhin** [comment]
@omesquita as a temporary solution (until the issue is fixed properly) you can provide a custom default configuration:
```kotlin
class MyGlobalConfigProvider: GlobalConfigProvider {
  override fun get() = Config.Builder()
    .setInstrumentedPackages(...)
    .build()
}
```
and inject it via java service configuration:
```
$ cat META-INF/services/org.robolectric.pluginapi.config.GlobalConfigProvider
my.package.MyGlobalConfigProvider
```
`RobolectricTestRunner` will automatically pick up your config.

Explanation: Robolectric uses Java `ServiceLoader` to resolve several pieces of configuration, including default config.
You can see what exactly is configured if you look into robolectric's.jar and check out `META-INF/services/`

**Comment by calvarez-ov** [comment]
For info, we still have this in 4.7.2, if we remove our `instrumentedPackages=androidx.loader.content` workaround

**Comment by hoisie** [comment]
The fix for this isn't in Robolectric, it's in espresso.

It was fixed in https://github.com/android/android-test/pull/1000

You may need to update to espresso 3.5 alpha.

**Comment by hoisie** [comment]
Here is an example of how to fix this issue:

https://github.com/robolectric/robolectric/pull/6875/files

Just update to `espresso-core >= 3.5.0-alpha01`

`testImplementation "androidx.test.espresso:espresso-core:3.5.0-alpha01"`

**Comment by hoisie** [comment]
@alwa Espresso needs to be updated, not AXT Core. This ModernAsyncTask logic is related to Espresso's IdlingResources.

**Comment by calvarez-ov** [comment]
I confirm, if I upgrade to espresso to `3.5.0-alpha03`, I can remove `instrumentedPackages=androidx.loader.content` from our `robolectric.properties`, and our tests pass!

👏

**Comment by omesquita** [comment]
> @omesquita as a temporary solution (until the issue is fixed properly) you can provide a custom default configuration:
> 
> ```kotlin
> class MyGlobalConfigProvider: GlobalConfigProvider {
>   override fun get() = Config.Builder()
>     .setInstrumentedPackages(...)
>     .build()
> }
> ```
> 
> and inject it via java service configuration:
> 
> ```
> $ cat META-INF/services/org.robolectric.pluginapi.config.GlobalConfigProvider
> my.package.MyGlobalConfigProvider
> ```
> 
> `RobolectricTestRunner` will automatically pick up your config.
> 
> Explanation: Robolectric uses Java `ServiceLoader` to resolve several pieces of configuration, including default config. You can see what exactly is configured if you look into robolectric's.jar and check out `META-INF/services/`

Thanks @nborodikhin! This solution worked to me.

**Comment by utzcoz** [comment]
Anyone has this problem can upgrade Espresso to 3.5.0-alpha03(https://github.com/robolectric/robolectric/issues/6593#issuecomment-975829808) and later version(3.5.0-alpha04 is latest now) to fix this problem if you can use alpha library, or you can follow https://github.com/robolectric/robolectric/issues/6593#issuecomment-971149643 as temporary solution to fix this problem before Espresso 3.5.0 stable released(this stable version not released yet, see https://developer.android.com/jetpack/androidx/releases/test#espresso_350_2). Feel free to reopen this issue if you want to discuss more about it.
